pytorch_lightning/callbacks/pruning.py

```diff
@@ -0,0 +1,396 @@
+            raise MisconfigurationException(
+                'Only "unstructured" PRUNING_TYPE supported for '
+                f"the `pruning_method`. Found method {pruning_fn} of type {pruning_fn.PRUNING_TYPE}. "
+            )
+
+        self.pruning_fn = pruning_fn
+
+        if not (prune_on_epoch_end or prune_on_fit_end):
+            raise MisconfigurationException(
+                "The `ModelPruning` won't be triggered as currently not activated on either `epoch_end` or `fit_end`."
+            )
+
+        self.make_pruning_permanent = make_pruning_permanent
+
+        if not isinstance(amount, (int, float, Callable)):
+            raise MisconfigurationException(
+                "amount should be provided and be either an int, a float or Callable function."
+            )
+
+        self.amount = amount
+
+    def filter_parameters_to_prune(self, parameters_to_prune: Optional[List[Tuple[nn.Module, str]]]):
+        """
+        This function can be overriden to control which module to prune.
+        """
+        return parameters_to_prune
+
+    def _create_pruning_fn(self, pruning_fn: str, *args, **kwargs):
+        """
+        This function takes `pruning_fn`, a function name.
+
+        IF use_global_unstructured, pruning_fn will be resolved into its associated ``PyTorch BasePruningMethod``
+        ELSE, pruning_fn will be resolved into its function counterpart from `torch.nn.utils.prune`.
+
+        """
+        if self.use_global_unstructured:
+            pruning_fn = _PYTORCH_PRUNING_METHOD[pruning_fn]
+            self._global_kwargs = kwargs
+            return pruning_fn
+        else:
+            return ModelPruning._wrap_pruning_fn(_PYTORCH_PRUNING_FUNCTIONS[pruning_fn], **kwargs)
+
+    @staticmethod
+    def _wrap_pruning_fn(pruning_fn, *args, **kwargs):
+        return partial(pruning_fn, **kwargs)
+
+    def _make_pruning_permanent(self):
+        for module, param_name in self.parameters_to_prune:
+            pytorch_prune.remove(module, param_name)
+
+    def _resolve_amount(self, current_epoch: int) -> float:
+        if isinstance(self.amount, Callable):
+            amount_fn = self.amount
+            amount = amount_fn(current_epoch)
+        else:
+            amount = self.amount
+        return amount
+
+    def _restore_original_weights(self, module: nn.Module, orig_module: nn.Module, tensor_name: str):
+        """
+        "The lottery ticket hypothesis" (https://arxiv.org/pdf/1803.03635.pdf) algorithm:
+
+            1. Randomly initialize a neural network f(x;θ0)(where θ0 ∼Dθ).
+            2. Train the network for j iterations, arriving at parameters θj .
+            3. Prune p% of the parameters in θj , creating a mask m.
+            4. Reset the remaining parameters to their values in θ0, creating the winning ticket f(x; m⊙θ0).
+
+        This function is responsible of step 4.
+        """
+        trained = getattr(module, tensor_name)
+        orig = getattr(orig_module, tensor_name)
```

Comment:
nit: I would encourage adding either asserts for missing attribute or defaults in the getattr. That will not only shush the future linters, but also provide some basic safety

Reply:
Done !